File: ruby/json_parse.rb
#!/usr/bin/env ruby

require "json"
require "benchmark"

puts 'hash to json'
3.times do |i|

  n = 1000 ** i
  h = {
    name: "bob jones",
    reference: "his reference"
  }
  beginning_time = Time.now
  time = Benchmark.measure do
    n.times do
      h.to_json
    end
  end

  end_time = Time.now
  elapsed = (end_time - beginning_time) * 1000.0
  each = elapsed / n
  printf(" %13.5f ms per query   | %9d iterations    | %8.5f ms total time\n", each, n, elapsed)

end

puts 'json to hash'
3.times do |i|
  n = 1000 ** i
  j = "{\"name\":\"bob jones\",\"reference\":\"his reference\"}"
  beginning_time = Time.now
  time = Benchmark.measure do
    n.times do
      JSON.parse(j)
    end
  end

  end_time = Time.now
  elapsed = (end_time - beginning_time) * 1000.0
  each = elapsed / n
  printf(" %13.5f ms per query   | %9d iterations    | %8.5f ms total time\n", each, n, elapsed)
end

File: go/json_parse.go
package main

import (
	"encoding/json"
	"fmt"
	"math"
	"time"
)

type Bench struct {
	Name      string `json:"name"`
	Reference string `json:"reference"`
}

func main() {
	fmt.Println("struct to json")
	struct_to_json()

	fmt.Println("json to struct")
	json_to_struct()
}

func struct_to_json() {

	bench := Bench{"bob jones", "his reference"}

	for k := 0; k < 3; k++ {
		n := int(math.Pow(1000, float64(k)))
		start := time.Now()
		for i := 0; i < n; i++ {
			json.Marshal(bench)
		}

		elapsed := time.Since(start)
		Results(elapsed, n)
	}
}

func json_to_struct() {

	j := []byte("{\"name\":\"bob jones\",\"reference\":\"his reference\"}")

	bench := &Bench{}

	for k := 0; k < 3; k++ {
		n := int(math.Pow(1000, float64(k)))

		start := time.Now()
		for i := 0; i < n; i++ {
			json.Unmarshal(j, &bench)
		}
		elapsed := time.Since(start)
		Results(elapsed, n)
	}

}

func Results(elapsed time.Duration, n int) {
	each := elapsed / time.Duration(n)
	fmt.Printf(" %13.5f ms per query   | %9d iterations    | %8.5f ms total time\n", milliseconds(each), n, milliseconds(elapsed))
}

func milliseconds(dur time.Duration) float64 {
	nanoseconds := float64(dur.Nanoseconds())
	milliseconds := nanoseconds / 1000000.0
	return milliseconds
}
